File: 0027. Remove Element/Solution.java
class Solution {
    // 双指针法
    // Approach 1: Two Pointers
    public int removeElement_MK1(int[] nums, int val) {
        int last = 0;
        for (int num : nums) {
            if (num != val) {
                nums[last++] = num;
            }
        }
        return last;
    }

    // 改进的双指针法，当要删除的元素很少时，速度更快
    // Approach 2: Two Pointers - when elements to remove are rare
    public int removeElement_MK2(int[] nums, int val) {
        int len = nums.length;
        for (int i = 0; i < len; ) {
            if (nums[i] == val) {
                nums[i] = nums[--len];
            } else {
                i++;
            }
        }
        return len;
    }
}

File: 0027. Remove Element/Solution.c
// 双指针法
// Approach 1: Two Pointers
int removeElement_MK1(int *nums, int numsSize, int val)
{
    int last = 0;
    for (int i = 0; i < numsSize; i++)
        if (nums[i] != val)
            nums[last++] = nums[i];
    return last;
}

// 改进的双指针法，当要删除的元素很少时，速度更快
// Approach 2: Two Pointers - when elements to remove are rare
int removeElement_MK2(int *nums, int numsSize, int val)
{
    int tail = numsSize;
    for (int i = 0; i < tail; )
        if (nums[i] == val)
            nums[i] = nums[--tail];
        else
            i++;
    return tail;
}
